Store bucket created time as a metadata

balamurugana opened the pull request:
Fixes #9459

## Description


## Motivation and Context


## How to test this PR?


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

balamurugana committed:
Store bucket created time as a metadata

Fixes #9459

klauspost commented on cmd/bucket-handlers.go:
```diff
@@ -48,6 +49,7 @@ const (
 	objectLockConfig                  = "object-lock.xml"
 	bucketObjectLockEnabledConfigFile = "object-lock-enabled.json"
 	bucketObjectLockEnabledConfig     = `{"x-amz-bucket-object-lock-enabled":true}`
+	bucketCreatedTimeFile             = "created.time"
```
Start with a `.`
```suggestion
	bucketCreatedTimeFile             = ".created.time"
```

klauspost commented on cmd/bucket-handlers.go:
```diff
@@ -292,6 +294,16 @@ func (api objectAPIHandlers) ListBucketsHandler(w http.ResponseWriter, r *http.R
 		}
+
+		for i := range bucketsInfo {
+			configFile := path.Join(bucketConfigPrefix, bucketsInfo[i].Name, bucketCreatedTimeFile)
```
This is a significant overhead, just by the IO alone. Do we have to read through this all the time?

kannappanr replied:
This is not expected to be called frequently like say list objects. This is supposed to be a not so frequent call.

klauspost commented on cmd/bucket-handlers.go:
```diff
@@ -292,6 +294,16 @@ func (api objectAPIHandlers) ListBucketsHandler(w http.ResponseWriter, r *http.R
+
+		for i := range bucketsInfo {
+			configFile := path.Join(bucketConfigPrefix, bucketsInfo[i].Name, bucketCreatedTimeFile)
+			if configData, err := readConfig(ctx, objectAPI, configFile); err == nil {
```
Let's make this structured, so we can modify the data later.

```
type bucketMetaData struct {
	Created time.Time
}
```

Since we are doing new [metadata storage](https://github.com/minio/minio/pull/9377) for objects, let's do it in a similar way for buckets.

You can check out how it is serialized by following this: https://github.com/minio/minio/pull/9377/files#diff-a80127ea76a22ee96c3d31240f99b42bR14

klauspost replied:
I see you are also responsible for  `bucketObjectLockEnabledConfigFile` which would have been nice to have in the same file as well. ;)

klauspost committed:
Add bucket metadata file (#16)

For now includes creation date and lock config as raw XML.

Should do rolling upgrades from existing lock config.

Not tested.

PR against #9465 / balamurugana:fix-issue-9459

balamurugana committed:
add bucket metadata only for created time

balamurugana committed:
migrate object-lock-enabled flag to bucket metadata

harshavardhana commented on cmd/bucket-meta.go:
```diff
@@ -0,0 +1,148 @@
+	legacyBucketObjectLockEnabledConfigFile = "object-lock-enabled.json"
+	legacyBucketObjectLockEnabledConfig     = `{"x-amz-bucket-object-lock-enabled":true}`
+
+	bucketMetadataFile    = ".metadata"
```
we can call this `.bucket.meta` or `.metadata.bin` to indicate its a binary file?

balamurugana replied:
done

harshavardhana commented on cmd/object-lock.go:
```diff
@@ -376,22 +376,15 @@ func checkPutObjectLockAllowed(ctx context.Context, r *http.Request, bucket, obj
 		ctx := logger.SetReqInfo(GlobalContext, &logger.ReqInfo{BucketName: bucket.Name})
-		configFile := path.Join(bucketConfigPrefix, bucket.Name, bucketObjectLockEnabledConfigFile)
-		bucketObjLockData, err := readConfig(ctx, objAPI, configFile)
+		meta, err := loadBucketMetadata(ctx, objAPI, bucket.Name)
```
loadBucketMetadata can return errConfigNotFound - we should ignore and proceed.

balamurugana replied:
If any error from `loadBucketMetadata()` means an error. This function also migrates existing object-lock-enabled configuration.

harshavardhana replied:
Correct but if there is no object lock enabled, this function returns errConfigNotFound is that expected?

balamurugana replied:
No. It doesn't. `errConfigNotFound` error means no bucket metadata found.

harshavardhana replied:
Oh I see now it's fixed..

harshavardhana replied:
> No. It doesn't. `errConfigNotFound` error means no bucket metadata found.

https://github.com/minio/minio/commit/119919167f440977cb819f7b1404dea6b8c66979 in this commit the change wasn't present, it was added later

https://github.com/minio/minio/commit/37c6231b1d9c44052c58152b4d2a83a65f57912a#diff-c4485f758c1bb2ceeebafaa7f537ac20R112

Now I see it all good!

klauspost commented on cmd/bucket-handlers.go:
```diff
@@ -292,6 +290,16 @@ func (api objectAPIHandlers) ListBucketsHandler(w http.ResponseWriter, r *http.R
+				bucketsInfo[i].Created = meta.Created
+			default:
+				logger.LogIf(ctx, err)
+			}
```
```suggestion
			meta, err := loadBucketMetadata(ctx, objectAPI, bucketsInfo[i].Name)
			logger.LogIf(ctx, err)
			if err == nil {
				bucketsInfo[i].Created = meta.Created
			}
```

balamurugana replied:
Done

klauspost commented on cmd/bucket-meta.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+// newBucketMetadata creates bucketMetadata with the supplied name and Created to Now.
+func newBucketMetadata(name string, lockEnabled bool) bucketMetadata {
```
```suggestion
func newBucketMetadata(name string) bucketMetadata {
```
Set this afterwards on the returned object. We don't want be have to add all content as we add more.

balamurugana replied:
Done

klauspost commented on cmd/bucket-meta.go:
```diff
@@ -0,0 +1,157 @@
+		return meta, err
+	}
+
+	// Control here means old bucket without bucket metadata. Hence we migrate existing settings.
```
Move this to a `func (b *bucketMetadata) convertLegacyLockconfig() error` or something like it.

This will make it clear what it does.

balamurugana replied:
Done

klauspost commented on cmd/bucket-meta.go:
```diff
@@ -0,0 +1,157 @@
+
+	if err != errConfigNotFound {
+		return meta, err
+	}
```
With your change we will initialize bucket creation time to 'Now'.

You might have missed that one of the [changes](https://github.com/balamurugana/minio/pull/16/files#diff-7d2b37ed440839b3878664247c692005R379) I proposed would transfer the time from the bucket listing, so that would be preserved.

Did that present a problem?

balamurugana replied:
Got the problem. Fixed it.

klauspost reviewed (commented):
I still don't understand why we need to read 2 files to figure out the locking configuration.

balamurugana committed:
addressed review comments

klauspost reviewed (approved):
LGTM.  `initBucketObjectLockConfig` should now convert everything on startup.

harshavardhana committed:
Merge branch 'master' into fix-issue-9459

kannappanr reviewed (approved):
LGTM
